File: packages/harness/src/unattended.ts
/**
 * What the agent may ask the host to do while nobody is watching.
 *
 * Two asks, and they are the same act: **the host performs a write the agent may
 * not, on the agent's request, decided host-side.** Landing a Worktree writes the
 * live tree; cutting a Pre-release writes the root manifest. Both paths are
 * closed to the agent by `denyWrite` and stay closed — see
 * docs/adr/0023-a-second-door-rather-than-a-wider-one.md, which argues the trade
 * once so neither half has to argue it again.
 *
 * ## Why this is beside ./preview.ts rather than inside it
 *
 * A Preview is a *launch*: nothing about the developer's clone changes, and the
 * host's whole share of the decision is "is this a name git reported". These two
 * change the clone. So they carry a second thing a Preview does not — a **gate**,
 * asked host-side against facts the agent cannot influence — and this module is
 * the confined half of both: the tool's shape, the sentences the agent reads
 * back, and the one line each request and answer crosses on.
 *
 * The gate itself is not here. It is `landWorktree` in ./landing.ts, which asks
 * `unattendedLanding` in ./fence.ts, and it runs in the Harness runtime where git
 * runs. Nothing in this file decides anything.
 *
 * ## The answers carry a sentence, and ./preview.ts's do not
 *
 * That is a deliberate departure and it is worth being exact about. A preview
 * answer is a **tag**, because the process that produces it is the Rust host —
 * the one holding the Credential and performing spawns — and a sentence composed
 * there is the string most likely to carry a path, an environment or an OS error
 * into a confined process.
 *
 * These answers are produced in the *runtime*, in TypeScript, by the same code
 * that performed the merge. That is exactly the division `report-merge` already
 * makes: `mergeBriefing` writes the sentence, the Rust host copies it onto the
 * pipe and composes nothing. So a `detail` here is a sentence some TypeScript
 * wrote about work it did, and the rule that matters is kept — **every outcome
 * the Rust host decides for itself arrives with no detail at all**, and its
 * sentence is authored below like a preview's. `src-tauri/src/unattended.rs`
 * asserts that.
 *
 * A refusal has to say *which rule refused and about what*, or a run report
 * prints "it would not land" and a developer wakes up to a branch and no reason.
 * That sentence is `unattendedLanding`'s and it is already written; carrying a
 * tag alone would mean writing it a second time on this side, out of fields, in
 * a process that never saw the branch.
 */

// ---------------------------------------------------------------------------
// Landing a Worktree
// ---------------------------------------------------------------------------

/** The Custom Tool's name, as the agent sees it. */
export const LAND_WORKTREE_TOOL = 'land_worktree'

/**
 * What the tool says it does — including what it will refuse and why.
 *
 * The refusals are described rather than left to be discovered, for the reason
 * `LAUNCH_PREVIEW_DESCRIPTION` describes its own: an agent that knows the Fence
 * is never landed unattended hands that branch over instead of calling this and
 * reading a refusal, and a tool call that was never made is the cheapest refusal
 * there is.
 *
 * The last paragraph is the one that stops a wrong Turn rather than a wasted
 * one. An agent that has merged its own branch is an agent whose next reasonable
 * thought is "my change is running", and it is not: varnick is still the build
 * from before, exactly as after a developer's merge (`RESTART_STILL_OWED`).
 */
export const LAND_WORKTREE_DESCRIPTION = [
  'Ask varnick to merge one of your worktrees into the live tree: squashed onto the live branch,',
  'the worktree removed and the branch deleted. This is how work you finished is delivered without',
  'a person awake to press anything.',
  '',
  'Takes the name of a worktree under .claude/worktrees/ — one path component, nothing else. It is',
  'not a path and not a command, and the merge is performed by the host, because a confined process',
  'cannot write the live tree.',
  '',
  'The host decides, from git rather than from anything you say. It reads what the branch changed',
  'out of git itself and refuses the whole branch if any of it is protected: packages/harness/**,',
  'src-tauri/**, the sandbox policy and its baseline, scripts/**, .githooks/**, .varnick/gitconfig,',
  'or a package.json diff that adds, changes or removes an install lifecycle script. A dirty live',
  'tree and a branch that does not merge cleanly are refused too. Every refusal comes back with the',
  'rule and the path that caused it, in words a run report can print — a refused branch is finished',
  'work waiting for a person, not a mistake to retry.',
  '',
  'A branch that lands is in the live tree and is still not running: varnick goes on serving the',
  'build it started with until it restarts. Do not assume the change is live because it merged.',
].join('\n')

/**
 * What happened to a request to land a Worktree.
 *
 * Seven, and they divide in three: one that landed, four the host refused about
 * *this* branch, and two where the question could not be put. The last two are
 * separated from the refusals on purpose — "protected, so a human merges it" is
 * a finished ticket to hand over, and "the host could not be asked" is a broken
 * machine to stop on. An orchestrator that read one for the other would either
 * park good work or retry a refusal all night.
 */
export const LANDING_OUTCOMES = [
  /** The branch is on the live tree. */
  'landed',
  /** The protected-path predicate said no. The detail says which rule and what. */
  'refused',
  /** The live tree has uncommitted work in it, so a merge could lose something. */
  'dirty-live-tree',
  /** The branch conflicts, or git could not say whether it merges. */
  'unmergeable',
  /** No worktree of that name has commits the live tree does not. */
  'unknown-worktree',
  /** git could not be asked what worktrees exist, so nothing was checked. */
  'no-worktrees',
  /** The question reached nothing, or the merge did not finish. Nothing landed. */
  'no-landing',
] as const

export type LandingOutcome = (typeof LANDING_OUTCOMES)[number]

export function isLandingOutcome(value: unknown): value is LandingOutcome {
  return typeof value === 'string' && (LANDING_OUTCOMES as readonly string[]).includes(value)
}

/**
 * What the host answered, as it crosses into the Sandbox.
 *
 * `detail` is a sentence composed where the work happened, or `null`. See the
 * module header for which outcomes may carry one and which may never.
 */
export interface UnattendedAnswer<Outcome extends string> {
  readonly outcome: Outcome
  readonly detail: string | null
}

export type LandingAnswer = UnattendedAnswer<LandingOutcome>

/** What to tell the agent, in one sentence, selected by the tag. */
export function landingOutcomeMessage(outcome: LandingOutcome): string {
  switch (outcome) {
    case 'landed':
      return 'That worktree is merged into the live tree. varnick is still running the build it started with, so the change is not live until it restarts.'
    case 'refused':
      return 'varnick will not land that branch without a human, and the rule that refused it is:'
    case 'dirty-live-tree':
      return 'The live tree has uncommitted work in it. A merge over that is how a change nobody knew about is lost, so nothing was merged. Nothing here can commit it for you — this one waits for the developer.'
    case 'unmergeable':
      return 'That branch does not merge into the live tree as it stands, so nothing was merged. Merge the live branch down into the worktree, where you may write, and ask again.'
    case 'unknown-worktree':
      return 'That is not the name of a worktree under .claude/worktrees/ with commits the live tree does not have. The name is one path component — not a path, not the live clone — and a branch level with the live tree has nothing to land.'
    case 'no-worktrees':
      return 'The host could not ask git what worktrees exist, so nothing was checked and nothing was merged.'
    case 'no-landing':
      return 'The host could not carry out the landing. Nothing was merged, and this is a broken machine rather than a refusal — do not retry it in a loop.'
  }
}

/**
 * The tool's answer, as the agent reads it.
 *
 * A sentence and a flag, like `previewToolResult`. The flag is what makes "your
 * branch is still pending" a fact the agent can act on rather than prose it has
 * to interpret: six of these seven mean nothing was merged, and an agent that
 * went on to report the ticket delivered would be reporting a merge that never
 * happened.
 *
 * The detail is appended rather than interpolated, so an outcome that arrives
 * with none still reads as a whole sentence.
 */
export function landingToolResult(answer: LandingAnswer): {
  readonly landed: boolean
  readonly text: string
} {
  return {
    landed: answer.outcome === 'landed',
    text: sentence(landingOutcomeMessage(answer.outcome), answer.detail),
  }
}

/**
 * One request to land a Worktree, as one line on the agent host's stdout.
 *
 * The third shape on a pipe that carries Turn events and Preview requests, told
 * apart the same way: by which id and which kind it names. Two fields, because
 * there is nothing else a landing needs — the host resolves the name against
 * `git worktree list`, and there is deliberately no field for a base, a commit,
 * a message or a path list. What the branch changed is read from git, and a
 * field the agent could fill would be a way to answer the predicate's question
 * for it.
 */
export function encodeLandingRequest(requestId: string, worktree: string): string {
  return `${JSON.stringify({ kind: 'land-worktree', requestId, worktree })}\n`
}

// ---------------------------------------------------------------------------
// Cutting a Pre-release
// ---------------------------------------------------------------------------

/** The Custom Tool's name, as the agent sees it. */
export const CUT_PRE_RELEASE_TOOL = 'cut_pre_release'

/**
 * What the tool says it does, and the one thing it deliberately does not.
 *
 * Promotion is the developer's and stays the developer's, so the description
 * says so: an agent that knows a Pre-release is not served does not go looking
 * for the control that would serve it, and does not tell the developer in its
 * report that the new build is running.
 */
export const CUT_PRE_RELEASE_DESCRIPTION = [
  'Cut a pre-release from the work that has landed: a version, a changelog entry, an announcement',
  'written from the run\'s tickets, a build and a tag. This is how a night of finished tickets ends',
  'with one thing the developer can accept in the morning.',
  '',
  'Takes the feature slug the run is named by — the one under .scratch/, as in .scratch/<slug>/issues.',
  'One path component, nothing else. The host runs the release, because cutting one writes the root',
  'manifest and package.json is not yours to write.',
  '',
  'It does not change what varnick is serving and it does not promote anything. The developer',
  'promotes, from a band in the window. A run that cut a pre-release ends there.',
].join('\n')

/**
 * What happened to a request to cut a Pre-release.
 *
 * The same three-way split the release script's own exit codes make, and it is
 * the split that matters to whatever is orchestrating: `cut` is a night
 * delivered, `refused` is a decision the release made on purpose — every ticket
 * parked, nothing landed to announce — and `no-release` is the world broken
 * underneath it. See `packages/core/scripts/release.ts`, which is where those
 * three are decided.
 */
export const RELEASE_OUTCOMES = [
  /** A pre-release exists, is tagged, and is not being served. */
  'cut',
  /** There was nothing to release, and the detail says what the release said. */
  'refused',
  /** The slug is not a feature slug. Nothing ran. */
  'not-a-feature',
  /** The release could not be run or did not answer. Nothing was cut. */
  'no-release',
] as const

export type ReleaseOutcome = (typeof RELEASE_OUTCOMES)[number]

export function isReleaseOutcome(value: unknown): value is ReleaseOutcome {
  return typeof value === 'string' && (RELEASE_OUTCOMES as readonly string[]).includes(value)
}

export type ReleaseAnswer = UnattendedAnswer<ReleaseOutcome>

/** What to tell the agent, in one sentence, selected by the tag. */
export function releaseOutcomeMessage(outcome: ReleaseOutcome): string {
  switch (outcome) {
    case 'cut':
      return 'A pre-release is cut, recorded and tagged. It is not being served: the developer promotes it, and until they do varnick goes on running the build it started with.'
    case 'refused':
      return 'Nothing was cut, and that is a decision the release made rather than a failure:'
    case 'not-a-feature':
      return 'That is not a feature slug. It is one path component naming the run — the directory under .scratch/ the tickets are in — and it is not a path, a flag or a version.'
    case 'no-release':
      return 'The release could not be run, so nothing was cut, tagged or announced. This is a broken machine rather than a refusal — do not retry it in a loop.'
  }
}

/**
 * The tool's answer, as the agent reads it.
 *
 * `cut` is the flag rather than "not refused", so the two ways of not cutting
 * stay apart at the one place the agent reads them: a run that reported a
 * pre-release because the tool did not fail would be a run that announced work
 * nobody can accept.
 */
export function releaseToolResult(answer: ReleaseAnswer): {
  readonly cut: boolean
  readonly text: string
} {
  return {
    cut: answer.outcome === 'cut',
    text: sentence(releaseOutcomeMessage(answer.outcome), answer.detail),
  }
}

/**
 * One request to cut a Pre-release, as one line on the agent host's stdout.
 *
 * One field beside the id, and it is the slug. There is no field for a version,
 * a tag, a changelog or an artifact: all four are decided by
 * `packages/core/release.ts` from what actually landed, and a request that could
 * name a version would be an agent announcing whatever it liked.
 */
export function encodeReleaseRequest(requestId: string, feature: string): string {
  return `${JSON.stringify({ kind: 'cut-release', requestId, feature })}\n`
}

/**
 * Whether a string is the slug of a run, in the one shape that may reach argv.
 *
 * Asked in the Harness runtime, which is the process that spawns
 * `bun run release <slug>` — the string becomes a command's argument there and
 * nowhere else, so that is where the shape is decided. The Rust host forwards it
 * unexamined, for the reason it forwards a worktree diff's path: a validation
 * written on both sides is a validation that drifts, and the side that must be
 * right is the side holding the spawn.
 *
 * It is `is_plain_worktree_name` in src-tauri/src/preview.rs by another name,
 * and the same three shapes it refuses are the ones that matter here: a
 * separator, so the slug cannot climb out of `.scratch/`; a leading `-`, so it
 * cannot be read as a flag by the script it is handed to; and a leading `.`, so
 * it is not `..` in disguise. `release.ts` refuses the first two itself, which is
 * a second answer rather than the only one — a script's own argument check is
 * one edit away from being relaxed by somebody who does not know an agent's
 * string reaches it.
 */
export function isFeatureSlug(value: string): boolean {
  return (
    value.length > 0 &&
    value.length <= 100 &&
    !value.startsWith('-') &&
    !value.startsWith('.') &&
    [...value].every(
      (character) =>
        (character >= 'a' && character <= 'z') ||
        (character >= 'A' && character <= 'Z') ||
        (character >= '0' && character <= '9') ||
        character === '-' ||
        character === '_' ||
        character === '.',
    )
  )
}

/**
 * The sentence for a tag, and the detail after it when there is one.
 *
 * Trimmed, because the details are composed from git output and from a script's
 * last line, and a trailing newline in the middle of a tool result is a sentence
 * that reads as two.
 */
function sentence(message: string, detail: string | null): string {
  const said = detail === null ? '' : detail.trim()
  return said.length === 0 ? message : `${message} ${said}`
}

File: src-tauri/src/unattended.rs
// What the agent may ask this host to do, host-side: land a branch, cut a
// pre-release.
//
// Both are writes to the developer's clone that the agent itself may not make —
// `git merge` writes `packages/core/**` and a release bumps `package.json`, and
// `denyWrite` refuses the agent both, in the live tree, on every branch and
// none. **That does not change here.** What this file adds is a second, narrower
// door: the host performs the write, on the agent's request, and only when the
// answer to a question the agent cannot influence says it may. See
// docs/adr/0023-a-second-door-rather-than-a-wider-one.md.
//
// ## This process decides almost nothing
//
// It is the middle of three, and that is the whole of its job:
//
//   * the **agent** names a worktree, or a feature slug. Nothing else crosses;
//   * **this process** turns a worktree name into a path, by looking it up in a
//     table `git worktree list` built — `resolve_worktree` in preview.rs, reused
//     rather than reimplemented, because a second lookup is a second answer;
//   * the **Harness runtime** asks `unattendedLanding` about what git says the
//     branch changed, and merges only if it may. That is where every decision
//     is, in TypeScript, provable with no repository at all.
//
// The feature slug is forwarded unexamined, which is the one place this differs
// from the worktree name and it is deliberate. It becomes an argument to
// `bun run release` in the runtime and nowhere else, so the runtime is where its
// shape is decided (`isFeatureSlug` in packages/harness/src/unattended.ts) — the
// same call bridge.rs already makes about a worktree diff's path, and for the
// reason written there: a validation on both sides is a validation that drifts,
// and the side that has to be right is the side holding the spawn.
//
// ## Prose crosses, and it is never this process's prose
//
// A Preview's answer is a tag and nothing else, because a sentence composed here
// is the string most likely to carry a path, an environment or an OS error into
// a confined process — this is the host that holds the Credential and performs
// spawns. That rule is kept, in the form that survives a refusal having to
// explain itself: **an outcome this process decides carries no detail at all**,
// and a detail exists only when it came up from the runtime, which wrote it in
// TypeScript about work it did. That is the division `report_merge` already
// makes for a merge briefing, and the tests below are written against it.

use serde_json::Value;

/// How the agent asks for a Worktree to be landed. One line on its stdout.
pub const LAND_WORKTREE_KIND: &str = "land-worktree";

/// How the agent asks for a Pre-release to be cut.
pub const CUT_RELEASE_KIND: &str = "cut-release";

/// How this host answers a landing. One control line on the agent's stdin.
pub const LANDING_ANSWER_KIND: &str = "landing-answer";

/// How it answers a release.
pub const RELEASE_ANSWER_KIND: &str = "release-answer";

/// The outcomes a landing may end in, as `LANDING_OUTCOMES` in
/// packages/harness/src/unattended.ts names them.
///
/// Mirrored rather than shared, like `PreviewOutcome::tag` beside it, and the
/// mirror is load-bearing rather than documentation: the confined half refuses a
/// control request whose outcome it does not know, so an answer forwarded with a
/// tag this build invented would leave a tool call waiting for ever. Anything
/// not on this list becomes {@link NO_LANDING} on the way past.
pub const LANDING_OUTCOMES: [&str; 8] = [
    "landed",
    "refused",
    "dirty-live-tree",
    "unmergeable",
    "branch-moved",
    "unknown-worktree",
    "no-worktrees",
    "no-landing",
];

/// The outcomes a release may end in. See {@link LANDING_OUTCOMES}.
pub const RELEASE_OUTCOMES: [&str; 4] = ["cut", "refused", "not-a-feature", "no-release"];

/// The name is not one git reports under `.claude/worktrees/`.
pub const UNKNOWN_WORKTREE: &str = "unknown-worktree";

/// git could not be asked what worktrees exist, so nothing was checked.
pub const NO_WORKTREES: &str = "no-worktrees";

/// The question reached nothing, or the answer was not one this build can read.
///
/// **Never `refused`.** A refusal is a decision about a branch, and an
/// orchestrator hands a refused branch to a person and stops — so a runtime that
/// did not answer must not produce one, or a night's finished work is reported
/// as fenced when nothing looked at it.
pub const NO_LANDING: &str = "no-landing";

/// The release could not be run, or did not answer readably.
pub const NO_RELEASE: &str = "no-release";

/// One request to land a Worktree, as the agent host wrote it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LandingRequest {
    /// Which call this answers. Chosen by the agent host, opaque here.
    pub request_id: String,
    /// What the agent asked for. **Not yet a path, and not yet trusted.**
    pub worktree: String,
}

/// One request to cut a Pre-release.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReleaseRequest {
    pub request_id: String,
    /// The run being released. Forwarded, never joined onto anything here.
    pub feature: String,
}

/// What happened, as it goes back to the agent.
///
/// The detail is a sentence the *runtime* composed, or nothing. See the module
/// header: this process never fills it in.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Answer {
    pub outcome: String,
    pub detail: Option<String>,
}

impl Answer {
    /// An outcome this process decided, which by construction carries no prose.
    pub fn tag(outcome: &str) -> Self {
        Answer {
            outcome: outcome.to_string(),
            detail: None,
        }
    }
}

/// A line the agent host wrote, if it is a request to land a Worktree.
///
/// Rebuilt out of two strings, like every other parse on this wire. A request
/// carrying a `path`, a `base` or a list of changed paths alongside loses them
/// here — which is what makes "the predicate is asked about what git says"
/// a property of the parse rather than a rule someone upstream keeps.
pub fn landing_request_of(line: &str) -> Option<LandingRequest> {
    let (request_id, worktree) = two_strings(line, LAND_WORKTREE_KIND, "worktree")?;
    Some(LandingRequest {
        request_id,
        worktree,
    })
}

/// A line the agent host wrote, if it is a request to cut a Pre-release.
///
/// One field beside the id, and it is the slug. A request naming a version, a
/// tag or an artifact loses all three here: what a release announces is decided
/// by `packages/core/release.ts` from what actually landed, and an agent that
/// could name a version would be an agent announcing whatever it liked.
pub fn release_request_of(line: &str) -> Option<ReleaseRequest> {
    let (request_id, feature) = two_strings(line, CUT_RELEASE_KIND, "feature")?;
    Some(ReleaseRequest {
        request_id,
        feature,
    })
}

/// The two strings one of these requests is, or nothing.
///
/// An empty request id is refused because it names no call, and an answer to no
/// call is an answer nothing resolves. The argument may be empty — that is a
/// name that resolves to nothing and a slug that is not one, both of which are
/// answers rather than parse failures, and both of which say so in words.
fn two_strings(line: &str, kind: &str, field: &str) -> Option<(String, String)> {
    let value: Value = serde_json::from_str(line.trim()).ok()?;
    if value.get("kind").and_then(Value::as_str)? != kind {
        return None;
    }
    let request_id = value.get("requestId").and_then(Value::as_str)?;
    let argument = value.get(field).and_then(Value::as_str)?;
    if request_id.is_empty() {
        return None;
    }
    Some((request_id.to_string(), argument.to_string()))
}

/// The answer, as one control line for the agent host's stdin.
///
/// `serde_json` escapes newlines, so nothing here can split a control request
/// across two lines — the same framing every other line on this pipe uses, and
/// it matters more for this one: a detail is composed from git's account of a
/// branch, which is text nobody here chose.
///
/// A detail that is absent is absent, rather than an empty string: the confined
/// half drops a blank one anyway, and sending `""` would mean this process had
/// said something.
pub fn answer_line(kind: &str, request_id: &str, answer: &Answer) -> String {
    let line = match &answer.detail {
        Some(detail) => serde_json::json!({
            "kind": kind,
            "requestId": request_id,
            "outcome": answer.outcome,
            "detail": detail,
        }),
        None => serde_json::json!({
            "kind": kind,
            "requestId": request_id,
            "outcome": answer.outcome,
        }),
    };
    format!("{line}\n")
}

/// What the runtime said, or the tag that says it did not say it readably.
///
/// **The one place a string from another process becomes an answer**, so it is
/// pure and it is where the tests are. Two rules:
///
///   * the outcome must be one this build knows, because the confined half
///     refuses one it does not and a refused control request is a tool call that
///     never returns;
///   * the detail is taken only when the outcome was, so a reply that carried
///     prose beside an unreadable tag loses the prose with it. Nothing this
///     process observed may become a sentence the agent reads, and the safest
///     reading of "I did not understand the answer" is that there is no answer.
pub fn answer_of(reply: &Value, known: &[&str], fallback: &'static str) -> Answer {
    let Some(outcome) = reply.get("outcome").and_then(Value::as_str) else {
        return Answer::tag(fallback);
    };
    if !known.contains(&outcome) {
        return Answer::tag(fallback);
    }
    let detail = reply
        .get("detail")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|said| !said.is_empty())
        .map(str::to_string);
    Answer {
        outcome: outcome.to_string(),
        detail,
    }
}

/// The whole of what happens when the agent asks for a Worktree to be landed.
///
/// Three steps, and the first two are this process's entire contribution: which
/// worktrees exist, and which of them the agent named. Neither is a decision
/// about the branch — those are all one hop further on, in the runtime, where
/// `unattendedLanding` is asked about what git says the branch changed.
///
/// The clone root is this process's own — `VARNICK_CLONE_ROOT` or the build path
/// — and never anything the agent sent, exactly as `answer_preview`'s is.
///
/// A runtime that will not answer is `no-landing`, which fails closed: nothing
/// is merged, and the tag says the machine rather than the branch. There is no
/// path here that merges anything without the runtime having said it may.
pub fn answer_landing(app: &tauri::AppHandle, worktree: &str) -> Answer {
    use tauri::Manager;

    let clone_root = crate::bridge::clone_root(std::env::var(crate::bridge::CLONE_ROOT_VAR).ok());

    let Some(listing) = crate::preview::worktree_listing(&clone_root) else {
        return Answer::tag(NO_WORKTREES);
    };
    // The same table a Preview is resolved against, and the same rule: the
    // agent's string is only ever a key, and the answer is a value git wrote.
    // `../..`, an absolute path and a name git does not report are one answer.
    let Ok(path) = crate::preview::resolve_worktree(worktree, &listing, &clone_root) else {
        return Answer::tag(UNKNOWN_WORKTREE);
    };

    match app.state::<crate::bridge::HarnessRuntime>().land_worktree(&path) {
        Ok(reply) => answer_of(&reply, &LANDING_OUTCOMES, NO_LANDING),
        // The failure is not forwarded. It is this process's account of a pipe,
        // and the agent's answer is a tag whose sentence is written on the far
        // side — see the module header.
        Err(_) => Answer::tag(NO_LANDING),
    }
}

/// The whole of what happens when the agent asks for a Pre-release.
///
/// One step, because there is nothing here to resolve: a slug names a directory
/// under `.scratch/` that the *release* reads, and this process neither joins it
/// onto anything nor passes it to a command. It goes to the runtime as a field,
/// which checks its shape before it can become an argument.
pub fn answer_release(app: &tauri::AppHandle, feature: &str) -> Answer {
    use tauri::Manager;

    match app
        .state::<crate::bridge::HarnessRuntime>()
        .cut_pre_release(feature)
    {
        Ok(reply) => answer_of(&reply, &RELEASE_OUTCOMES, NO_RELEASE),
        Err(_) => Answer::tag(NO_RELEASE),
    }
}

#[cfg(test)]
mod tests {
    use super::{
        answer_line, answer_of, landing_request_of, release_request_of, Answer, LandingRequest,
        ReleaseRequest, LANDING_OUTCOMES, NO_LANDING, NO_RELEASE, RELEASE_OUTCOMES,
    };
    use serde_json::json;

    // -----------------------------------------------------------------------
    // What may cross, on the way out
    // -----------------------------------------------------------------------

    #[test]
    fn a_landing_request_is_two_strings_and_a_changed_path_list_is_not_one_of_them() {
        /*
          The failure this parse exists to prevent. A request that brought its
          own account of what the branch changed would be the agent answering
          the protected-path question about itself — so the fields are rebuilt,
          and everything else sent alongside is a field that was never read.
        */
        let request = landing_request_of(
            r#"{"kind":"land-worktree","requestId":"l1","worktree":"agent-one",
                "changed":["README.md"],"base":"HEAD","force":true,"path":"/etc"}"#,
        );
        assert_eq!(
            request,
            Some(LandingRequest {
                request_id: "l1".to_string(),
                worktree: "agent-one".to_string(),
            })
        );
    }

    #[test]
    fn a_release_request_cannot_name_a_version_a_tag_or_an_artifact() {
        let request = release_request_of(
            r#"{"kind":"cut-release","requestId":"r1","feature":"autonomous-runs",
                "version":"9.9.9","tag":"v9.9.9","artifact":"/tmp/x.tgz"}"#,
        );
        assert_eq!(
            request,
            Some(ReleaseRequest {
                request_id: "r1".to_string(),
                feature: "autonomous-runs".to_string(),
            })
        );
    }

    #[test]
    fn the_three_shapes_on_this_pipe_are_told_apart_and_none_reads_another() {
        // A Turn event, a Preview request and these two share one pipe and are
        // told apart by kind and by which id they name. A parse that accepted a
        // neighbour would deliver a landing to a Preview, or a Turn to a merge.
        let landing = r#"{"kind":"land-worktree","requestId":"l1","worktree":"agent-one"}"#;
        let release = r#"{"kind":"cut-release","requestId":"r1","feature":"runs"}"#;

        assert!(landing_request_of(release).is_none());
        assert!(release_request_of(landing).is_none());
        assert!(landing_request_of(r#"{"kind":"delta","turnId":"t1","text":"hello"}"#).is_none());
        assert!(release_request_of(r#"{"kind":"delta","turnId":"t1","text":"hello"}"#).is_none());
        assert!(crate::preview::preview_request_of(landing).is_none());
        assert!(crate::agent::agent_event_of(landing).is_none());
        assert!(crate::agent::agent_event_of(release).is_none());
    }

    #[test]
    fn a_request_missing_a_half_is_not_a_request() {
        for broken in [
            r#"{"kind":"land-worktree"}"#,
            r#"{"kind":"land-worktree","requestId":"l1"}"#,
            r#"{"kind":"land-worktree","worktree":"agent-one"}"#,
            r#"{"kind":"land-worktree","requestId":"","worktree":"agent-one"}"#,
            r#"{"kind":"land-worktree","requestId":"l1","worktree":42}"#,
            "not json",
            r#"{"ready":true}"#,
        ] {
            assert!(landing_request_of(broken).is_none(), "{broken}");
        }
        for broken in [
            r#"{"kind":"cut-release"}"#,
            r#"{"kind":"cut-release","requestId":"r1"}"#,
            r#"{"kind":"cut-release","feature":"runs"}"#,
            r#"{"kind":"cut-release","requestId":"","feature":"runs"}"#,
        ] {
            assert!(release_request_of(broken).is_none(), "{broken}");
        }
    }

    #[test]
    fn a_name_that_resolves_to_nothing_is_still_a_request_because_the_answer_is_a_sentence() {
        // Refusing to parse it would leave the tool call unanswered. It parses,
        // and `answer_landing` turns it into `unknown-worktree`, which the agent
        // reads as a sentence saying what a worktree name is.
        assert_eq!(
            landing_request_of(r#"{"kind":"land-worktree","requestId":"l1","worktree":"../.."}"#)
                .map(|request| request.worktree),
            Some("../..".to_string())
        );
    }

    // -----------------------------------------------------------------------
    // What may cross, on the way back
    // -----------------------------------------------------------------------

    #[test]
    fn an_answer_is_exactly_one_line_and_carries_the_runtimes_sentence_untouched() {
        let line = answer_line(
            super::LANDING_ANSWER_KIND,
            "l1",
            &Answer {
                outcome: "refused".to_string(),
                detail: Some(
                    "src-tauri/src/bridge.rs is protected — src-tauri/** may not be landed without a human."
                        .to_string(),
                ),
            },
        );
        assert!(line.ends_with('\n'));
        assert_eq!(line.matches('\n').count(), 1);
        assert_eq!(
            serde_json::from_str::<serde_json::Value>(line.trim()).unwrap(),
            json!({
                "kind": "landing-answer",
                "requestId": "l1",
                "outcome": "refused",
                "detail": "src-tauri/src/bridge.rs is protected — src-tauri/** may not be landed without a human.",
            })
        );
    }

    #[test]
    fn an_outcome_this_host_decided_carries_no_prose_at_all() {
        /*
          The rule the whole module is written to keep. This process holds the
          Credential and performs spawns; a sentence composed here is the string
          most likely to carry a path or an environment into the Sandbox. So
          every answer it decides for itself is a bare tag, and the sentences for
          those tags are authored in packages/harness/src/unattended.ts.
        */
        for outcome in [
            super::UNKNOWN_WORKTREE,
            super::NO_WORKTREES,
            NO_LANDING,
            NO_RELEASE,
        ] {
            let answer = Answer::tag(outcome);
            assert_eq!(answer.detail, None);
            let parsed: serde_json::Value =
                serde_json::from_str(answer_line("landing-answer", "l1", &answer).trim()).unwrap();
            assert!(parsed.get("detail").is_none(), "{outcome} said something");
        }
    }

    #[test]
    fn an_outcome_this_build_does_not_know_becomes_the_tag_that_says_nothing_happened() {
        /*
          A runtime one version ahead — or a reply that lost its way — must not
          become an answer nothing wrote. The confined half checks the tag
          against its own closed list and refuses a control request that fails
          it, and a refused control request is a tool call that never returns.
          So an unknown tag is converted here, into the one outcome that is
          honest about it.

          And it must not be `refused`: that reads as a decision about the
          branch, which nobody made.
        */
        for reply in [
            json!({ "outcome": "landed?" }),
            json!({ "outcome": "declined" }),
            json!({ "outcome": 7 }),
            json!({ "detail": "it went fine" }),
            json!({}),
        ] {
            let answer = answer_of(&reply, &LANDING_OUTCOMES, NO_LANDING);
            assert_eq!(answer.outcome, NO_LANDING);
            assert_eq!(answer.detail, None, "prose survived an unreadable tag");
        }
        assert_eq!(
            answer_of(&json!({ "outcome": "cut?" }), &RELEASE_OUTCOMES, NO_RELEASE).outcome,
            NO_RELEASE
        );
    }

    #[test]
    fn every_outcome_the_runtime_can_send_survives_the_crossing() {
        // The mirror of `LANDING_OUTCOMES` and `RELEASE_OUTCOMES` in
        // packages/harness/src/unattended.ts. A tag missing from either list
        // here is a tool call that hangs, so the list is asserted whole rather
        // than sampled.
        for outcome in LANDING_OUTCOMES {
            let reply = json!({ "outcome": outcome, "detail": "because." });
            let answer = answer_of(&reply, &LANDING_OUTCOMES, NO_LANDING);
            assert_eq!(answer.outcome, outcome);
            assert_eq!(answer.detail, Some("because.".to_string()));
        }
        for outcome in RELEASE_OUTCOMES {
            assert_eq!(
                answer_of(&json!({ "outcome": outcome }), &RELEASE_OUTCOMES, NO_RELEASE).outcome,
                outcome
            );
        }
        // The two lists are not each other's, which is what keeps a release
        // answer from reading as a landing.
        assert_eq!(
            answer_of(&json!({ "outcome": "landed" }), &RELEASE_OUTCOMES, NO_RELEASE).outcome,
            NO_RELEASE
        );
    }

    #[test]
    fn a_blank_detail_is_no_detail_rather_than_a_blank_sentence() {
        for reply in [
            json!({ "outcome": "landed", "detail": "" }),
            json!({ "outcome": "landed", "detail": "   " }),
            json!({ "outcome": "landed", "detail": 42 }),
            json!({ "outcome": "landed" }),
        ] {
            assert_eq!(answer_of(&reply, &LANDING_OUTCOMES, NO_LANDING).detail, None);
        }
    }
}
